### PR TITLE
feat(object_detector): add require_motion_overlap to gate recordings on object/motion overlap

### LINKED ISSUE #1363: Feature request: correlate motion contours with detected objects before triggering recordings
Hi,

I would like to suggest an improvement to how motion detection and object detection can be combined for event recording.

Current behavior:

I use MOG2 motion detection together with YOLO object detection, with `scan_on_motion_only: true` and `require_motion: true` on some object labels such as `car`.

This works well in principle, but there is one common false-positive case:

- moving tree shadows trigger MOG2 motion somewhere in the frame
- YOLO detects a parked car elsewhere in the frame
- since motion is currently detected globally, the parked car can trigger an event recording
- however, the car itself did not move

From what I understand, `require_motion` currently means “there is motion somewhere in the frame”, not “there is motion overlapping this detected object”.

Proposed improvement:

It would be very useful to have an option to require spatial correlation between the detected object and the motion detector result before starting an event recording.

For example, per label:

```yaml
labels:
  - label: car
    trigger_event_recording: true
    require_motion: true
    require_motion_overlap: true
    motion_overlap_threshold: 0.1
```

Possible behavior:

compute the object bounding box
compute the MOG2/background-subtractor motion contours or bounding boxes
only allow the object to trigger recording if the object box overlaps a motion contour / motion bounding box
optionally allow a configurable margin around the object bbox
optionally expose an overlap threshold
This would help with parked cars, parked bikes, garden furniture, etc., where unrelated motion such as shadows, leaves, rain, or reflections should not validate a static object.

Why this matters
For outdoor cameras, motion caused by shadows or trees is very common. Masking is not always possible because the parked object can be in different places. Disabling the object label entirely is also not ideal, because moving cars should still be recorded.

A spatial “motion near/inside the object” condition would make require_motion much more useful for this type of setup.

Feature request: show live motion tuning values in the camera fine-tuning UI
It would also be very helpful if the camera fine-tuning page could show live motion detector debug values directly in the UI.

For example:

current max_area
configured area threshold
number of detected contours
largest contour bounding box
whether motion is currently considered active
maybe a small recent history graph
optionally an overlay of motion contours on the camera frame
This would make it much easier to tune area, threshold, history, and learning_rate without having to watch container logs.

Example useful debug output:

text

camera=driveway
motion=clear
max_area=0.00064
threshold=0.4
contours=1
largest_bbox=(0.897, 0.54, 0.937, 0.577)
Having these values visible directly in the fine-tuning camera window would make motion tuning much more practical, especially for outdoor cameras with changing light, shadows, rain, and vegetation.

Thanks for considering this.

### PR DESCRIPTION
**Pull request by mvanhorn**
## Summary

Adds a per-label `require_motion_overlap` option (plus `motion_overlap_threshold`) so an object only triggers/keeps an event recording when its bounding box spatially overlaps a motion contour, rather than when there is motion anywhere in the frame.

## Why this matters

Today `require_motion` only means "there is motion somewhere in the frame," not "there is motion overlapping this detected object." As reported in #1363, with `scan_on_motion_only: true` and `require_motion: true` on a label such as `car`, a parked car can still trigger an event recording: MOG2 motion fires on moving tree shadows in one region while YOLO detects the stationary car elsewhere, producing a false recording. Tying the motion requirement to the object's own bounding box removes those false triggers while leaving the existing whole-frame `require_motion` behavior intact for users who want it.

## Changes

- New per-label config options in `viseron/domains/object_detector/const.py` and `LABEL_SCHEMA`:
  - `require_motion_overlap` (bool, default `False`)
  - `motion_overlap_threshold` (float 0-1, default `0.1`) - the minimum fraction of the object's bounding box that must be covered by motion contours.
- New geometry helper `helpers.object_motion_overlap(rel_bbox, rel_contours)` that rasterizes the relative motion contours and returns the fraction of the object's bbox they cover (deterministic, resolution-independent, unit-tested).
- `Filter` stores the two options and exposes `require_motion_overlap` / `motion_overlap_threshold` properties.
- `nvr.py` consults the overlap helper at both the start gate (`trigger_event_recording`) and the stop gate (`event_over_check_motion`) when `require_motion_overlap` is set, via `_object_has_motion_overlap`. The motion scanner is also kept active (at startup and after a recording stops) whenever a label needs motion overlap, so motion contours keep being produced even when the motion detector itself does not trigger recordings.

## Backward compatibility

`require_motion_overlap` defaults to `False`, so existing `require_motion` behavior is unchanged. When no motion detector is configured, the overlap requirement is disabled (mirroring the existing `require_motion` handling).

## Tests

- Geometry helper unit tests (full overlap, disjoint, partial, empty contours, zero-area bbox).
- Schema defaults/round-trip for the two new options.
- NVR test verifying recording does not start when motion exists but does not overlap the object, and starts once the motion contour overlaps it.

Closes #1363

### CODE REVIEW

**Comment by john-** [comment]
The objective is to not have repeat  detections for stationary objects?  If so how does this compare to a TBD approach that compares bounding box/class on subsequent frames and discards the detection that match?  For example, car bounding box does not change in next frame so object is considered stationary?

Of course I am over simplifying this and there may be considerations that make it not viable.  If feasible, this would not require the correlation between motion and object detections and therefore no need for motion detection for this purpose.

**Comment by roflcoopter** [comment]
Thank you! Docs needs to be re-generated to pass CI.

The Copilot review left some interesting notes. What do you think, do they make any sense?

**Comment by roflcoopter** [comment]
> The objective is to not have repeat detections for stationary objects? If so how does this compare to a TBD approach that compares bounding box/class on subsequent frames and discards the detection that match? For example, car bounding box does not change in next frame so object is considered stationary?
> 
> Of course I am over simplifying this and there may be considerations that make it not viable. If feasible, this would not require the correlation between motion and object detections and therefore no need for motion detection for this purpose.

This is not quite the same thing. While this will help to reduce false positives and also reduce to problem of stationary objects, an object tracking model in addition to this is the best way to make sure objects are tracked properly between frames.

**Comment by john-** [comment]
> This is not quite the same thing. While this will help to reduce false positives and also reduce to problem of stationary objects, an object tracking model in addition to this is the best way to make sure objects are tracked properly between frames.

Thanks for explaining this to me.  I am looking forward to the enhancement.

**Comment by mvanhorn** [comment]
Docs are regenerated and pushed, the generate_docs check is green now. On the Copilot notes, I went through them against the code. Short version: one is minor, three point at something real.

1. Start-vs-stop description (`const.py`): partly fair. `event_over_check_motion` does route through `_object_has_motion_overlap`, so overlap refines the stop check too when `require_motion` is on, not just the start gate. "require_motion still controls stopping" is still true as the master switch, but I can reword so it doesn't imply overlap is start-only.

2. `require_motion_overlap` without `require_motion`: valid. The start gate (`trigger_event_recording`) fires on `require_motion or require_motion_overlap`, but the stop gate (`event_over_check_motion`) only checks `require_motion`. So a label with `require_motion_overlap: true` and `require_motion: false` starts on overlap but motion never stops it. Either the stop gate should honor `require_motion_overlap` too, or overlap should be treated as a modifier that only applies when `require_motion` is set (and warn on the standalone combo, like the existing no-motion-detector guard). I lean toward coupling the stop gate so start and stop stay symmetric.

3. No contours blocks recording: also fair as a defensive case. `_object_has_motion_overlap` returns False when `motion_detected` is true but `motion_contours` is empty, so a detector reporting motion without contours would silently block recordings. The built-in contour-based detectors always provide contours so it may not bite today, but a whole-frame fallback (motion detected + no contours = treat as overlapping) would be safer and cheap.

4. 1000x1000 mask per call: agreed, this is the one I'd prioritize. The mask only depends on the frame's motion contours, not the object, but it's rebuilt per object per frame in both gates. Building it once per frame (keyed on the contours) and reusing it for every object, or rasterizing only the bbox region, removes the repeated 1MB alloc + fillPoly on the hot path.

Happy to fold 2, 3, and 4 into this PR with tests, or split them into a follow-up if you'd rather keep this one tight. Your call on scope.

**Comment by roflcoopter** [comment]
> 2\. `require_motion_overlap` without `require_motion`: valid. The start gate (`trigger_event_recording`) fires on `require_motion or require_motion_overlap`, but the stop gate (`event_over_check_motion`) only checks `require_motion`. So a label with `require_motion_overlap: true` and `require_motion: false` starts on overlap but motion never stops it. Either the stop gate should honor `require_motion_overlap` too, or overlap should be treated as a modifier that only applies when `require_motion` is set (and warn on the standalone combo, like the existing no-motion-detector guard). I lean toward coupling the stop gate so start and stop stay symmetric.

I agree, making the start and stop symmetric feels best.



> 3\. No contours blocks recording: also fair as a defensive case. `_object_has_motion_overlap` returns False when `motion_detected` is true but `motion_contours` is empty, so a detector reporting motion without contours would silently block recordings. The built-in contour-based detectors always provide contours so it may not bite today, but a whole-frame fallback (motion detected + no contours = treat as overlapping) would be safer and cheap.

Also agree. I recently introduced external motion detectors (triggered from MQTT for instance) which does not use contours, so using `require_motion_overlap` should be a no-op for these kinds of detectors.


> 4\. 1000x1000 mask per call: agreed, this is the one I'd prioritize. The mask only depends on the frame's motion contours, not the object, but it's rebuilt per object per frame in both gates. Building it once per frame (keyed on the contours) and reusing it for every object, or rasterizing only the bbox region, removes the repeated 1MB alloc + fillPoly on the hot path.

Do whatever feels best, maybe reducing the mask to something lower than 1000x1000 is a good idea as well? 


> Happy to fold 2, 3, and 4 into this PR with tests, or split them into a follow-up if you'd rather keep this one tight. Your call on scope.

Feel free to update this PR, good to keep everything organized 👍

Thanks for taking your time to improve Viseron!

**Comment by mvanhorn** [comment]
Pushed all three in a5929bb. The stop gate (event_over_check_motion) now fires on require_motion or require_motion_overlap, so start and stop stay symmetric. For the no-contours case, _object_has_motion_overlap returns True when motion is detected but there are no contours, so require_motion_overlap is a no-op for your MQTT-style external detectors. And the mask is now built once per frame and cached, keyed on the scaled contours, rather than rebuilt per object, with a test asserting fillPoly runs a single time across multiple objects. Ruff and the behavior smokes pass locally; I could not run the full pytest suite in my environment (pytest-postgresql was not available), so I am leaning on CI for the official run.

**Comment by roflcoopter** [comment]
Thanks for the quick update! I am preparing a release at the moment, will look through and merge this one after that is done

**Comment by roflcoopter** [comment]
Finally got the time to do a last review. Everything looks good from what i can see.

Will merge as soon as the conflicts are resolved, do you have time to fix that @mvanhorn ?

**Comment by mvanhorn** [comment]
Rebased onto `dev` (7e9e34d) and force-pushed as f21e24f. GitHub shows the PR mergeable again.

Two conflicts, both import-only:

- `viseron/components/nvr/nvr.py`: dev added `from viseron.helpers.validators import UNDEFINED` on the same line this branch had extended the `viseron.helpers` import with `object_motion_overlap`. Kept both.
- `tests/helpers/test__init__.py`: dev added `Mock, patch` and the `zoneinfo` imports where this branch had added `MagicMock`. Unioned them.

No behavior changes in the rebase. Verified locally on Python 3.12 (matching DEFAULT_PYTHON in ci.yaml): 150 tests pass across `tests/helpers/test__init__.py`, `tests/helpers/test_filter.py`, `tests/domains/object_detector/test_abstract_object_detector.py`, and `tests/components/nvr/test_nvr.py`, and flake8 and ruff are clean on the changed files. CI is running now.

**Comment by roflcoopter** [comment]
Perfect, thank you for this!
